Multi-variant repository analysis docs update (#13424)
github/codeql: docs/codeql/codeql-for-visual-studio-code/running-codeql-queries-at-scale-with-mrva.rst

You can then insert the ``new-repo-list`` of repositories into your list of custom repository lists for easy access in the Variant Analysis Repositories panel.

> Using code search to add repositories to a custom list 

Reviewer: I think it might be worth being explicit here that we're talking about searching GitHub.com.

Author: I updated to make it "GitHub code search" so its a bit more explicit, but hesitate to put GitHub.com because it could be mistaken for using the code search UI on GitHub.com to find and add repositories, whereas the main advantage of this feature is being able to use search directly in the extension. "Using GitHub.com code search in the extension to add repositories to a custom list" is probably most accurate but quite long

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> You can use code search directly in the CodeQL extension to add a subset of repositories from GitHub.com to a custom list. Note that this feature uses the legacy code search via the code search API. For more information on the syntax to use, see "`Searching code (legacy) <https://docs.github.com/en/search-github/searching-on-github/searching-code>`__."

Reviewer: I'm wondering if we want to put the note into a "Note"? If you like the idea Sarita, you can find an example on lines 111-115 in this file.

Author: yep good idea, will do!

> For example, to add all repositories in the ``rails`` organization on GitHub, you can search ``org:rails``. 

Author: @norascheuch @jf205 hello! I was wondering if there is a certain example search that might be helpful to add here? I did a very basic one but not sure if there is a name of a certain function or search term that might be helpful to illustrate the benefits. Thanks!

Reviewer 2: I think using this as an example is fine! I have been testing with some library names, but I don't know enough about the ecosystem to know which ones would be a better idea to share in the docs. (I used `SQLAlchemy` and `RuntimeExec` for example)

Custom lists can contain a maximum of 1000 repositories, so at most only the first 1000 repositories returned from your search will be added to your list.

#. In the Variant Analysis Repositories panel, choose the list that you want to add repositories to. You can create a new list or choose an existing list that already contains repositories.

#. Right-click on the list you have chosen and then click **Add repositories with GitHub Code Search**. 

#. In the search bar, select a language for your search from the choices in the dropdown. 

    .. image:: ../images/codeql-for-visual-studio-code/variant-analysis-code-search-language.png
        :alt: Screenshot of the search bar for using code search to add repositories to a custom list. The search bar asks you to choose a language for your search and has a dropdown list of languages to choose from.

> #. Type the search query that you want to use and press **Enter**.

Reviewer: I'm wondering if we also want a screenshot of this? I'm having trouble imagining it, but I expect it makes sense if you're using the extension 🙂 

Maybe more like:
```suggestion
#. In the XXX field, type the search query that you want to use and press **Enter**.
```

Author: I like your wording suggestion! I think it is pretty straightforward - just a search bar like appears in the screenshot above, but with no dropdown, so we can probably hold off on the screenshot!

#. You can view the progress of your search in the bottom right corner of the application in a box with the text "Searching for repositories...". If you click **Cancel**, no repositories will be added to your list.

> #. Once complete, you will see the resulting repositories appear in the dropdown under your custom list in the Variant Analysis Repositories panel.

Reviewer: These aren't really steps. Maybe convert them into a paragraph?

> It is possible that some of the resulting repositories do not have CodeQL databases or are not accessible. When you run an analysis on the list, the results view will show you which repositories could not be analyzed.

Reviewer: I think it would be helpful here to know how these are reported in the results view.

I suggest that we also reframe this slightly as I suspect that it will be normal to find some repositories without CodeQL databases.

I'm also wondering what would cause a repository to be accessible to your search, but where the CodeQL repository wasn't accessible. 🤔 @norascheuch - can you help with this?

Author: Good points! @norascheuch had mentioned in the docs-content issue that access could have been denied to the repository when the analysis is run, but it would be great to get more details if we want to add them to the docs!

I hesitated to go into further detail because from what I could tell, this is not necessarily specific to this feature, it is specific to MRVA generally and the results view of an analysis, which I believe would be covered in this section above: https://codeql.github.com/docs/codeql-for-visual-studio-code/running-codeql-queries-at-scale-with-mrva/#exploring-your-results. We could just link to that section at the end of this paragraph, since it does show a results view, though not specifically repositories that weren't analyzed.

Reviewer: In that case, maybe we could say what the results view would report?

Author: Updated with a little more info as far as I know (the repositories are broken into three categories in the results view - those that were analyzed, those that denied access, and those that had no CodeQL database

Reviewer 2: Sounds very good, thanks @saritai!

Reviewer 3: > I'm also wondering what would cause a repository to be accessible to your search, but where the CodeQL repository wasn't accessible

Not sure if it's worth going into detail in these docs, but to answer your question I'll try to list all the ways a repo might end up not being analysed.
- If the repo doesn't have a CodeQL database. This is already mentioned. There is a proposal for how we could teach the search API about CodeQL databases, but right now there's no guarantee that repos in the search results have CodeQL databases.
- If the repo is private and you don't have access to it. Though this shouldn't happen much if at all when the results came from the a search, because the search API will have already filtered the list to repos you can see. It's possible if repo visibility changes between running the search and running the MRVA.
- If the repo is private but the controller repo is public. For safety reasons we don't allow analysing private code in the public controller repo. The user could fix this themselves by adding `is:public` to their search. Sadly I don't think we could add this ourselves since at the point of the search we don't know if the controller repo is public or private.

Author: Thanks so much for the details! I might hold off on adding them to the docs now, but we can always add more detail later if it turns out to be something we think should be there!

You can then insert the ``new-repo-list`` of repositories into your list of custom repository lists for easy access in the Variant Analysis Repositories panel.

> Using GitHub code search to add repositories to a custom list 
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: This is probably giving the test failure - I forget this all the time:

```suggestion
Using GitHub code search to add repositories to a custom list 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

You can use code search directly in the CodeQL extension to add a subset of repositories from GitHub.com to a custom list. 

.. pull-quote::

    Note

    This feature uses the legacy code search via the code search API. For more information on the syntax to use, see "`Searching code (legacy) <https://docs.github.com/en/search-github/searching-on-github/searching-code>`__."

For example, to add all repositories in the ``rails`` organization on GitHub, you can search ``org:rails``. 

> Custom lists can contain a maximum of 1000 repositories, so at most only the first 1000 repositories returned from your search will be added to your list.

Reviewer 2: We decided last week to remove this limit of 1000 entries for a list. However, the api returns a maximum of 1000 results per search request. Therefore I think it makes sense to say that a maximum of 1000 repos can be added per search.

Author: Ahh great to know! Will update the language based on the above

#. Right-click on the list you have chosen and then click **Add repositories with GitHub Code Search**. 

#. In the pop-up that appears at the top of the application, under the search bar, select a language for your search from the choices in the dropdown. 

    .. image:: ../images/codeql-for-visual-studio-code/variant-analysis-code-search-language.png
        :alt: Screenshot of the search bar for using code search to add repositories to a custom list. The search bar asks you to choose a language for your search and has a dropdown list of languages to choose from.

#. In the search bar, type the search query that you want to use and press **Enter**.

You can view the progress of your search in the bottom right corner of the application in a box with the text "Searching for repositories...". If you click **Cancel**, no repositories will be added to your list. Once complete, you will see the resulting repositories appear in the dropdown under your custom list in the Variant Analysis Repositories panel.

> Not all resulting repositories will have CodeQL databases or allow access to be analyzed. When you run an analysis on the list, the Variant Analysis Results view will show you which repositories were analyzed, which denied access, and which had no CodeQL database.

Reviewer: Thanks for this addition. This is much clearer to users. I think we could be more explicit about the access, so maybe:

```suggestion
Some of the resulting repositories will not have CodeQL databases and some may not allow access by the CodeQL extension for Visual Studio Code. When you run an analysis on the list, the Variant Analysis Results view will show you which repositories were analyzed, which denied access, and which had no CodeQL database.
```
